### ISSUE TITLE
graceful rejection of invalid trades

### ISSUE THREAD

**Issue by dpaiton** [delvtech/elf-simulations#57]
Presently, the market class receives a proposed trade from a trader in a given block, and then attempts to execute the trade. If the trade was invalid (e.g. the user tries to buy more base than are available), the code breaks with assertion errors in the pricing model due to the math resulting in invalid outputs (such as a negative or complex-valued fee). 

Instead, the market should check if the trade is valid and reject it if the trade is not valid. The user is informed their trade was rejected (wallet deltas are zero, maybe some flag is set or something? RL users might like this information), and then the action & user loops continue.

**Comment by wakamex** [comment]
track here adding testing of individual market trades, full or partial open/close long/short. also some compound testing of the above.

**Comment by dpaiton** [comment]
market trade reverting:

market should identify trades that put it into an invalid sate, or one that validates agent stipulations (e.g. min_slippage). When it identifies an invalid trade, it raises an InvalidTrade exception. The simulator has a try/catch that catches these exceptions & cancels the trade.

if invalid trade is raised, we might still calculate gas fees & update states accordingly

**Comment by dpaiton** [comment]
An attempt was made here: https://github.com/element-fi/elf-simulations/blob/5fbf1be69ec9ebf3132e35bb4698f5e78ba89efd/src/elfpy/markets.py#L386

although we think there is a cleaner solution.

**Comment by dpaiton** [comment]
As an example:
This test:
```
self.setup_logging()
config = simulators.Config()
config.num_trading_days = 4
config.num_blocks_per_day = 4
config.variable_apr = [0.01] * config.num_trading_days
config.target_liquidity = 1_000_000
simulator = sim_utils.get_simulator(
    config=config,
    agents=[
        sim_utils.get_policy("single_long")(wallet_address=address, budget=1_000_000) for address in range(1, 3)
    ],
)
simulator.run_simulation()
```

fails:
```
  File "tests/test_sim.py", line 188, in <module>
    test.test_aggregate_agent_and_market_states()
  File "tests/test_sim.py", line 182, in test_aggregate_agent_and_market_states
    simulator.run_simulation()
  File "/Users/dylan/code/element/elf-simulations/elfpy/simulators/simulators.py", line 624, in run_simulation
    self.collect_and_execute_trades(liquidate)
  File "/Users/dylan/code/element/elf-simulations/elfpy/simulators/simulators.py", line 483, in collect_and_execute_trades
    self.execute_trades(trades)
  File "/Users/dylan/code/element/elf-simulations/elfpy/simulators/simulators.py", line 524, in execute_trades
    self.market.update_market(market_deltas)
  File "/Users/dylan/code/element/elf-simulations/elfpy/markets/hyperdrive.py", line 357, in update_market
    self.market_state.apply_delta(market_deltas)
  File "/Users/dylan/code/element/elf-simulations/elfpy/markets/hyperdrive.py", line 142, in apply_delta
    assert (
AssertionError: MarketState values must be > -1e-08. Error on bond_reserves = -38.423413853306556
```

because the single_long agent is opening big ass trades (note budget vs target liquidity). This is an example of where the market should gracefully reject trades that are too large. It shouldn't just throw an assertion error after the trades wrap up.

**Comment by dpaiton** [comment]
migrating related issue #241 from @villanuevawill to here:

In the following line of code, this is incorrect. Since the trade amount is supposed to be in base, slippage should allow for much larger values.

https://github.com/element-fi/elf-simulations/blob/main/elfpy/markets/hyperdrive.py#L527

**Comment by dpaiton** [comment]
Partially addressed by #331

**Comment by wakamex** [comment]
minor change implemented in #579 since I was just solving a case of zero trade amounts. this is a known case we want to have in our safety checks, so I added it. but there may be others we want to add later on, so not closing this issue.

**Comment by dpaiton** [comment]
I'm going to mark this as resolved & create a new issue for slippage controls. If we come up with additional checks for invalid trades we can make new issues for them.
